Note: Try to hover on the plots beacause this is interactive visualization.

If you are not able to see the images of plot in this notebook then Please visit:
http://nbviewer.jupyter.org/github/cyberdhiman/Assignment-Solution/blob/master/analyticsvidhya_Assignment%20%282%29.ipynb

The reason is use of Plotly library for visualization.

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/av_internship.csv


In [2]:
#reading files
import pandas as pd
av_data = pd.read_csv("../input/av_internship.csv")

In [45]:
#header files
from __future__ import absolute_import, division, print_function, unicode_literals
import numpy as np 
import pandas as pd
import plotly.express as px
import plotly.offline as pyo
import plotly.graph_objects as go
pyo.init_notebook_mode()

In [4]:
#data 
av_data

employee_id         department     region         education gender  \
0            65438  Sales & Marketing   region_7  Master's & above      f   
1            65141         Operations  region_22        Bachelor's      m   
2             7513  Sales & Marketing  region_19        Bachelor's      m   
3             2542  Sales & Marketing  region_23        Bachelor's      m   
4            48945         Technology  region_26        Bachelor's      m   
...            ...                ...        ...               ...    ...   
54803         3030         Technology  region_14        Bachelor's      m   
54804        74592         Operations  region_27  Master's & above      f   
54805        13918          Analytics   region_1        Bachelor's      m   
54806        13614  Sales & Marketing   region_9               NaN      m   
54807        51526                 HR  region_22        Bachelor's      m   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
0                sourcing                1   35                   5.0   
1                   other                1   30                   5.0   
2                sourcing                1   34                   3.0   
3                   other                2   39                   1.0   
4                   other                1   45                   3.0   
...                   ...              ...  ...                   ...   
54803            sourcing                1   48                   3.0   
54804               other                1   37                   2.0   
54805               other                1   27                   5.0   
54806            sourcing                1   29                   1.0   
54807               other                1   27                   1.0   

       length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                      8              1            0                  49   
1                      4              0            0                  60   
2                      7              0            0                  50   
3                     10              0            0                  50   
4                      2              0            0                  73   
...                  ...            ...          ...                 ...   
54803                 17              0            0                  78   
54804                  6              0            0                  56   
54805                  3              1            0                  79   
54806                  2              0            0                  45   
54807                  5              0            0                  49   

       is_promoted  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
54803            0  
54804            0  
54805            0  
54806            0  
54807            0  

[54808 rows x 14 columns]

In [5]:
#data shape
print("[Info]: Dataset Shape: ", av_data.shape)

[Info]: Dataset Shape:  (54808, 14)


In [6]:
#showing missing values
print("\n Missing data summary for the dataset:\n",av_data.isnull().sum())


 Missing data summary for the dataset:
 employee_id                0
department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64


For the education attribute 2409 observations and for previous_year_rating attribute 4124 are missing out of 54808 observations. 

In [7]:
#Missing Values Cleaning 
def clean(av_data):
    print("\n\n[OUT]: Cleaning Starts:")
    av_datac=av_data[av_data['education'].notnull() & av_data['previous_year_rating'].notnull()]
    print("\n[OUT]: Cleaning Done.")
    return av_datac
av_datac=clean(av_data)




[OUT]: Cleaning Starts:

[OUT]: Cleaning Done.


In [8]:

print("\n[Info]:Missing data summary for the dataset afte null value removal:\n",av_datac.isnull().sum())


[Info]:Missing data summary for the dataset afte null value removal:
 employee_id             0
department              0
region                  0
education               0
gender                  0
recruitment_channel     0
no_of_trainings         0
age                     0
previous_year_rating    0
length_of_service       0
KPIs_met >80%           0
awards_won?             0
avg_training_score      0
is_promoted             0
dtype: int64


In [9]:
print("[Info]: Dataset Shape after removal of observations which had missing values: ", av_datac.shape)

[Info]: Dataset Shape after removal of observations which had missing values:  (48660, 14)


In [10]:
#unique departments
departments=np.unique(av_datac['department'])

In [11]:
print("[Info]:Total departments: ",len(departments))
print("       Names of the departments:\n      ",list(departments))

[Info]:Total departments:  9
       Names of the departments:
       ['Analytics', 'Finance', 'HR', 'Legal', 'Operations', 'Procurement', 'R&D', 'Sales & Marketing', 'Technology']


In [12]:
#department wise employee
distribution_department=av_datac['department'].value_counts().rename_axis('Department Name').reset_index(name='Count of Employees')
fig = px.bar(distribution_department, x='Department Name', y='Count of Employees',title='Distribution of Data for Department',color='Department Name')
fig.show()

From the plot, it is clear that highest number of observations  is related to Sales and Marketing department and lowest to R&D department in the dataset.

In [13]:
#region wise employee
regions=av_datac['region'].unique()
print("[Info]: Total Number of regions: ",len(regions))
print(" List of regions:\n",list(regions))

[Info]: Total Number of regions:  34
 List of regions:
 ['region_7', 'region_22', 'region_19', 'region_23', 'region_26', 'region_2', 'region_20', 'region_34', 'region_1', 'region_4', 'region_29', 'region_31', 'region_15', 'region_14', 'region_11', 'region_5', 'region_28', 'region_17', 'region_13', 'region_16', 'region_25', 'region_10', 'region_27', 'region_30', 'region_12', 'region_21', 'region_32', 'region_6', 'region_33', 'region_8', 'region_24', 'region_3', 'region_9', 'region_18']


In [14]:
dbn_regions=av_datac['region'].value_counts().rename_axis('Regions').reset_index(name='Count of Employees')
fig = px.bar(dbn_regions, x='Regions', y='Count of Employees',title='Distribution of Regions',color="Count of Employees")
fig.show()

From plot, it is clear that, very large number of data is from region 2 and lowest is from region 18.

In [15]:

education=av_datac['education'].unique()
print("[Info]: Total Number of different education category: ",len(education))
print(" Education Category List:\n",list(education))

[Info]: Total Number of different education category:  3
 Education Category List:
 ["Master's & above", "Bachelor's", 'Below Secondary']


In [16]:
#education wise employees
dbn_education=av_datac['education'].value_counts().rename_axis('Education').reset_index(name='Count of Employees')
fig = px.bar(dbn_education, x='Education', y='Count of Employees',title='Distribution of data according to Education',color="Education")
fig.show()

It is clear from above plot that, very high number of employee have only Bachelor's degree.  

In [17]:
#gender based employees
dbn_gender=av_datac['gender'].value_counts().rename_axis('Gender').reset_index(name='Count')
fig = px.bar(dbn_gender, x='Gender', y='Count',title='Distribution of Employee according to Gender',color="Gender")
fig.show()


For above plot : Number of female employees is less than Male.

In [18]:
#Employees according to Channels
dbn_recr_channel=av_datac['recruitment_channel'].value_counts().rename_axis('Recruitment Channel').reset_index(name='Count')
fig = px.bar(dbn_recr_channel, x='Recruitment Channel', y='Count',title='Recruitment of Employees according to different Channel',color="Recruitment Channel")
fig.show()


From the above plot, it is clear that number of employees recruited by refferal is low and for majority of employees recruitment channel is not known.

In [19]:
#trainings done by employees
dbn_trainings=av_datac['no_of_trainings'].value_counts().rename_axis('Number of Trainings').reset_index(name='Count of Employees')
dbn_trainings=dbn_trainings.astype({'Number of Trainings':str})
fig = px.bar(dbn_trainings, x='Number of Trainings', y='Count of Employees',title='Number of trainings done by Employees',color="Count of Employees")
fig.show()

From above plot: Most of the employees have done only one training.

In [20]:
#max min age
print("\n[Info]: Maximum age of employee: ",av_data['age'].max())
print("\n[Info]: Minimum age of employee: ",av_data['age'].min())




[Info]: Maximum age of employee:  60

[Info]: Minimum age of employee:  20


In [21]:
#converting age to age groups
bins = [18,20, 25, 30,35, 40, 45,50, np.inf]
age_group=['18-20','20-25','25-30','30-35','35-40','40-45','45-50','50+']
av_datac['age_group']=pd.cut(av_datac['age'], bins, labels=age_group)


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [22]:
print("[Info]: Age group summary: ")
av_datac['age_group']

[Info]: Age group summary: 


0        30-35
1        25-30
2        30-35
3        35-40
4        40-45
         ...  
54802    30-35
54803    45-50
54804    35-40
54805    25-30
54807    25-30
Name: age_group, Length: 48660, dtype: category
Categories (8, object): [18-20 < 20-25 < 25-30 < 30-35 < 35-40 < 40-45 < 45-50 < 50+]

In [23]:
#employees distribution according to the age groups
dbn_ageGroup=av_datac['age_group'].value_counts().rename_axis('Age Group').reset_index(name='Count of Employees')
fig = px.bar(dbn_ageGroup, x='Age Group', y='Count of Employees',title='Age Group of Employees',color="Age Group")
fig.show()

Young employees are very very very less. Maximum number of employees belongs to 30-35 age group.

In [24]:
#employees ratings
dbn_p_ratings=av_datac['previous_year_rating'].value_counts().rename_axis('Rating Number').reset_index(name='Count of Employees')
dbn_p_ratings=dbn_p_ratings.astype({'Rating Number':str})
fig = px.bar(dbn_p_ratings, x='Rating Number', y='Count of Employees',title='Previous year ratings of Employees',color="Rating Number")
fig.show()

In [25]:
#eployees' service length
dbn_lservice=av_datac['length_of_service'].value_counts().rename_axis('Length of Service').reset_index(name='Count of Employees')
dbn_lservice=dbn_lservice.astype({'Length of Service':str})
fig = px.bar(dbn_lservice, x='Length of Service', y='Count of Employees',title='Number of Years for which a emlpoyee has served',color="Count of Employees")
fig.show()

From the above plot, It can be concluded that employees prefer to serve for 2 to 7 years.

In [26]:
#employess with KPI >80%
dbn_KPI=av_datac['KPIs_met >80%'].value_counts().rename_axis('KPIs_met >80%').reset_index(name='Count of Employees')
dbn_KPI=dbn_KPI.replace(to_replace=[0,1],value=["NO","Yes"])
fig = px.bar(dbn_KPI, x='KPIs_met >80%', y='Count of Employees',title='Whether KPIs met is greater than 80%',color="KPIs_met >80%")
fig.show()


Number of employees with KPI_met greater than 80% is less than than KPI_met less than 80%.

In [27]:
#number of employees which won award or not
dbn_award=av_datac['awards_won?'].value_counts().rename_axis('Award Won').reset_index(name='Count of Employees')
dbn_award=dbn_award.replace(to_replace=[0,1],value=["NO","Yes"])
fig = px.bar(dbn_award, x='Award Won', y='Count of Employees',title="Whether Employee won Award or not.",color="Award Won")
fig.show()


From the plot it is clear that very very less number of Employees won award.

In [28]:
#max min average trainings score 
print("\n[Info]: Maximum score of employee in training: ",av_data['avg_training_score'].max())
print("\n[Info]: Minimum score of employee in training ",av_data['avg_training_score'].min())


[Info]: Maximum score of employee in training:  99

[Info]: Minimum score of employee in training  39


In [29]:
#converting trainings score to group of ranges
bins = [0,30, 40, 50,60, 70, 80,90, np.inf]
score_range=['0-30','30-40','40-50','50-60','60-70','70-80','80-90','90+']
av_datac['training_score_range']=pd.cut(av_datac['avg_training_score'], bins, labels=score_range)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [30]:
#showing distribution of employees based on score of training
dbn_scoreRange=av_datac['training_score_range'].value_counts().rename_axis('Training Score Range').reset_index(name='Count of Employees')
fig = px.bar(dbn_scoreRange, x='Training Score Range', y='Count of Employees',title='Training Score of Employees w.r.t different range of score',color="Training Score Range")
fig.show()

Average training score range is 50-60.

In [31]:
#how many got promoted
dbn_promoted=av_datac['is_promoted'].value_counts().rename_axis('Promoted').reset_index(name='Count of Employees')
dbn_promoted=dbn_promoted.replace(to_replace=[0,1],value=["NO","Yes"])
fig = px.bar(dbn_promoted, x='Promoted', y='Count of Employees',title="Whether Employee was promoted or not.",color="Promoted")
fig.show()



Promotion was given only to few employees.

**Now Bivariate Exploration**

**Dependent and Independent**

In [32]:
#replacing numerical values of 0,1 to No and Yes for is_promoted
av_datac['is_promoted']=av_datac['is_promoted'].replace(to_replace=[0,1],value=["No","Yes"])
                                

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [33]:
#showing relation between department and promotion
department_and_Promoted=av_datac.groupby(["department","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(department_and_Promoted, x='department', y='Employee Count',title='The number of Employees  w.r.t Department and ispromoted',color='is_promoted',barmode="group")
fig.show()

department_and_Promoted=av_datac[av_datac['is_promoted']=='Yes'].groupby(["department"])['is_promoted'].aggregate("count").reset_index()
department_and_Promoted.columns=['Department','Count of Employee']
fig = px.bar(department_and_Promoted, x='Department', y='Count of Employee',title='The number of Employees which got promoted w.r.t Department',color="Department")
fig.show()



Employees from sales & marketing  and Operations are likely to get promotion.

In [34]:
#showing relation between Regions of employees and promotion
regions_and_Promoted=av_datac.groupby(["region","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
regions_and_Promoted=regions_and_Promoted.sort_values(by='Employee Count')
fig = px.bar(regions_and_Promoted, x='region', y='Employee Count',title='The number of Employees  w.r.t region and ispromoted',color='is_promoted',barmode="group")
fig.show()

Employees from region 2 are more likely to get promoted.

In [35]:
#showing relation between education of employees and promotion
education_and_Promoted=av_datac.groupby(["education","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(education_and_Promoted, x='education', y='Employee Count',title='The number of Employees  w.r.t education and ispromoted',color='is_promoted',barmode="group")
fig.show()

In [36]:
#showing realtion between gender of employees and promotion
gender_and_Promoted=av_datac.groupby(["gender","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
gender_and_Promoted=gender_and_Promoted.sort_values(by='Employee Count')
fig = px.bar(gender_and_Promoted, x='gender', y='Employee Count',title='The number of Employees  w.r.t Gender and ispromoted',color='is_promoted',barmode="group")
fig.show()

More male employees are promoted then female.

In [37]:

#showing realtion between recruiting channel of employees and promotion
rChannel_and_Promoted=av_datac.groupby(["recruitment_channel","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(rChannel_and_Promoted, x='recruitment_channel', y='Employee Count',title='The number of Employees which got promoted w.r.t Recruitment Channel',color='is_promoted',barmode="group")
fig.show()


In [38]:
#showing realtion between number of trainings done by employees and promotion
Trainings_and_Promoted=av_datac.groupby(["no_of_trainings","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')

fig = px.bar(Trainings_and_Promoted, x='no_of_trainings', y='Employee Count',title='The number of Employees w.r.t number of trainings and is_Promoted',color='is_promoted',barmode="group")
fig.show()


Employees with only one training are more likely to get promoted.

In [39]:
#showing realtion between age of  employees and promotion
age_and_Promoted=av_datac.groupby(["age","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')

fig = px.bar(age_and_Promoted, x='age', y='Employee Count',title='The number of Employees w.r.t number of age and is_Promoted',color='is_promoted',barmode="group")
fig.show()

This distribution looks like normal distribution.

In [40]:
#showing realtion between previous year ratings of employees and promotion
pyearRating_and_Promoted=av_datac.groupby(["previous_year_rating","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(pyearRating_and_Promoted, x='previous_year_rating', y='Employee Count',title='The number of Employees w.r.t previous year ratings and is_Promoted',color='is_promoted',barmode="group")
fig.show()

In [41]:

lservice_and_Promoted=av_datac.groupby(["length_of_service","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(lservice_and_Promoted, x='length_of_service', y='Employee Count',title='The number of Employees w.r.t length of service and is_Promoted',color='is_promoted',barmode="group")
fig.show()

Employees with less length of service are more likely to get promoted.

In [42]:
#howing realtion between KPI_met of employees and promotion
KPI_and_Promoted=av_datac.groupby(["KPIs_met >80%","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(KPI_and_Promoted, x='KPIs_met >80%', y='Employee Count',title='The number of Employees w.r.t KPI met > 80% and is_Promoted',color='is_promoted',barmode="group")
fig.show()

In [43]:
#showing realtion between awards won by employees and promotion
awards_and_Promoted=av_datac.groupby(["awards_won?","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(awards_and_Promoted, x='awards_won?', y='Employee Count',title='The number of Employees w.r.t award won and is_Promoted',color='is_promoted',barmode="group")
fig.show()

We can see that, it is very likely for  employee to get promotion if he/she has won an award.

In [44]:

#showing realtion between awards won by employees and promotion
trainingscore_and_Promoted=av_datac.groupby(["avg_training_score","is_promoted"])['employee_id'].count().reset_index(name='Employee Count')
fig = px.bar(trainingscore_and_Promoted, x='avg_training_score', y='Employee Count',title='The number of Employees w.r.t average training score and is_Promoted',color='is_promoted',barmode="group")
fig.show()

This is interesting plot, as individual peaks can be seen as part of different normal distribution.

I was not able to visualize more because My end semester exams are going on so I was not able to give much time.
So Please, I request you to consider my solution.